src/client/java/minicraft/entity/rideable/Boat.java

```diff
@@ -0,0 +1,292 @@
+					playerInBoat.render(screen);
+					break;
+
+				case NONE:
+					break;
+			}
+		} else {
+			screen.render(xo - 4, yo - 4, boatSprites[0][0]);
+
+		}
+	}
+
+	@Override
+	public void tick() {
+		tickTime++;
+
+		// moves the furniture in the correct direction.
+		move(pushDir.getX(), pushDir.getY());
+		pushDir = Direction.NONE;
+
+		if (pushTime > 0) {
+			pushTime--; // update pushTime by subtracting 1.
+		}
+
+		if (playerInBoat != null) { //If the player is in the boat.
+			exitTimer--;
+
+			if (exitTimer <= 0 && Game.input.getMappedKey("ATTACK").isDown()) { //Exit the boat
+				restorePlayer(playerInBoat);
+				return;
+			}
+
+			double ya = 0;
+			double xa = 0;
+
+			if (Game.input.getMappedKey("MOVE-UP").isDown()) ya -= 1;
+			if (Game.input.getMappedKey("MOVE-DOWN").isDown()) ya += 1;
+			if (Game.input.getMappedKey("MOVE-LEFT").isDown()) xa -= 1;
+			if (Game.input.getMappedKey("MOVE-RIGHT").isDown()) xa += 1;
+
+			move(xa, ya);
+			playerInBoat.x = x;
+			playerInBoat.y = y;
+			playerInBoat.stamina = Player.maxStamina;
+		}
+	}
+
+	@Override
+	public void hurt(Mob mob, int damage) {
+		if (mob instanceof Player) {
+			((Player) mob).isRiding = true;
+			playerInBoat = ((Player) mob);
+			exitTimer = 10;
+			playerInBoat.moveSpeed = 1.5;
+		}
+	}
+
+	@Override
+	public boolean blocks(Entity entity) {
+		return true;
+	}
+
+    @Override
+    public boolean interact(Player player, @Nullable Item item, Direction attackDir) {
+		//Put whatever item the player is holding into their inventory
+		if (player.activeItem != null && !(player.activeItem instanceof PowerGloveItem)) {
+			player.getInventory().add(0, player.activeItem);
+		}
+
+		if (item instanceof ToolItem && playerInBoat == null) {
+			level.dropItem(x, y, Items.get("Boat"));
+			remove();
+			return true;
+		} else {
+			player.isRiding = true;
+			playerInBoat = player;
+			exitTimer = 10;
+			playerInBoat.moveSpeed = 1.5;
+			return true;
+		}
+	}
+
+	public boolean move(double xa, double ya) {
+        if (Updater.saving || (xa == 0 && ya == 0)) {
+            return true; // pretend that it kept moving
+        }
+
+        // used to check if the entity has BEEN stopped, COMPLETELY; below checks for a lack of collision.
+        boolean stopped = true;
+
+        if (move2(xa, 0)) stopped = false; // becomes false if horizontal movement was successful.
+        if (move2(0, ya)) stopped = false; // becomes false if vertical movement was successful.
+
+        if (!stopped) {
+            int xt = x >> 4; // the x tile coordinate that the entity is standing on.
+            int yt = y >> 4; // the y tile coordinate that the entity is standing on.
+
+            // Calls the steppedOn() method in a tile's class. (used for tiles like sand (footprints) or lava (burning))
+            level.getTile(xt, yt).steppedOn(level, xt, yt, this);
+        }
+
+        return !stopped;
+    }
+
+    public boolean move2(double xa, double ya) {
```

Comment:
`move` and `move2` should probably use the corresponding methods in the `Entity` class.

Reply:
I’ll trt